### PR TITLE
Devices

### PR DESCRIPTION
**Pull request by fepitre**

### CODE REVIEW

File: _modules/ext_module_qvm.py [hunk 1]
```diff
@@ -941,8 +941,197 @@ def prefs(vmname, *varargs, **kwargs):
...
+        '--detach',
+        nargs='*',
+        default=[],
+        help='List of devices to detach'
+    )
+
+    args = qvm.parse_args(vmname, *varargs, **kwargs)
+
+    current_devices = []
+    for device_type in ['pci', 'bridge']:
```
Comment (marmarek) [inline _modules/ext_module_qvm.py]: There is also `block`, `usb`, and `mic`.

File: _modules/ext_module_qvm.py [hunk 2]
```diff
@@ -941,8 +941,197 @@ def prefs(vmname, *varargs, **kwargs):
...
+
+    # Helping function for parsing devices in args.attach and args.detach
+    def parse_device(raw_dev):
+        """
+        Example:
+            args.attach = [{u'bridge:sys-bridge:bridge0': [{u'ip': u'192.168.0.1'}, {u'netmask': u'255.255.255.0'}]}]
+        """
+        inline_dev = list(raw_dev.keys())[0]
+        device = {'device_type': inline_dev.split(':')[0], 'backend': inline_dev.split(':')[1],
+                  'dev_id': inline_dev.split(':')[2], 'options': {}}
```
Comment (marmarek) [inline _modules/ext_module_qvm.py]: Some meaningful error message about not enough `:`?

File: _modules/ext_module_qvm.py [hunk 3]
```diff
@@ -941,8 +941,197 @@ def prefs(vmname, *varargs, **kwargs):
...
+                    # args.vm.devices[device_type].update_persistent(a, False)
+                    args.vm.devices[device_type].detach(current_assignment)
+                    msg_options = '(' + ', '.join(
+                        '{}={}'.format(key, value) for key, value in current_assignment.options.items()) + ')'
+                    message_old = '[ATTACHED] ' + msg_options
+                break
+
+        try:
+            if device_type == 'pci' and ('no-strict-reset' not in device['options']):
+                device['options']['no-strict-reset'] = True
```
Comment (marmarek) [inline _modules/ext_module_qvm.py]: Why? Adding 'no-strict-reset' by default for every PCI device definitely is _not_ a good idea.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: _modules/ext_module_qvm.py (hunk 3 above)
Comment (fepitre) [reply _modules/ext_module_qvm.py]: Yes, it's a misunderstood of https://github.com/QubesOS/qubes-mgmt-salt-dom0-qvm/blob/e3c2c967290bf36c50ddb694edd69b257ca258cf/_modules/ext_module_qvm.py#L913
Up to the negation of the condition, I was thinking it was the default case.

File: _modules/ext_module_qvm.py (hunk 1 above)
Comment (fepitre) [reply _modules/ext_module_qvm.py]: Yes, so with respect to yesterday comments, no way of listing all the existing devices types?

File: _modules/ext_module_qvm.py (hunk 2 above)
Comment (fepitre) [reply _modules/ext_module_qvm.py]: Sure.

File: _modules/ext_module_qvm.py (hunk 1 above)
Comment (marmarek) [reply _modules/ext_module_qvm.py]: Unfortunately not, yet.

**Review by marmarek — APPROVED** [review]

File: _modules/ext_module_qvm.py [hunk 4]
```diff
@@ -941,8 +941,203 @@ def prefs(vmname, *varargs, **kwargs):
...
+        nargs='*',
+        default=[],
+        help='List of devices to detach'
+    )
+
+    args = qvm.parse_args(vmname, *varargs, **kwargs)
+
+    current_devices = []
+    for device_type in ['pci', 'block', 'usb', 'bridge', 'mic']:
+        for device in args.vm.devices[device_type].assignments():
```
Comment (marmarek) [inline _modules/ext_module_qvm.py:1029]: Enumerating devices of unknown type (for example if relevant addon isn't installed in dom0) will fail. Please wrap it in try/except for each device type. Or better, implement https://github.com/QubesOS/qubes-issues/issues/5213 first and properly enumerate device types.

File: _modules/ext_module_qvm.py [hunk 5]
```diff
@@ -941,8 +941,203 @@ def prefs(vmname, *varargs, **kwargs):
...
+
+    if args.detach:
+        args.detach = [parse_device(raw_dev) for raw_dev in json.loads(args.detach[0])]
+
+    for device in args.attach:
+        device_type = device['device_type']
+
+        if device['options'].get('pci_strictreset', False):
+            device['options']['no-strict-reset'] = not device['options']['pci_strictreset']
+            del device['options']['pci_strictreset']
```
Comment (marmarek) [inline _modules/ext_module_qvm.py]: `pci_strictreset` is a legacy VM option, not device option. It needs to stay in `prefs` when dealing with PCI devices for compatibility reasons, but no need to add it to this new code.

File: _modules/ext_module_qvm.py [hunk 6]
```diff
@@ -941,8 +941,203 @@ def prefs(vmname, *varargs, **kwargs):
...
+
+        message_old = None
+        for a in args.vm.devices[device_type].assignments():
+            if a.ident == device['dev_id']:
```
Comment (marmarek) [inline _modules/ext_module_qvm.py]: You need to compare backend too.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: _modules/ext_module_qvm.py (hunk 4 above)
Comment (fepitre) [reply _modules/ext_module_qvm.py:1029]: I will implement it.

**Comment by fepitre** [comment]
@marmarek: should be good now :)

**Review by marmarek — APPROVED** [review]

**Pull request merged**